tests/app/notifications/test_notifications_letter_callbacks.py

```diff
@@ -102,3 +104,186 @@ def _sample_sns_s3_callback(filename):
+        (
+            {"id": None},
+            "id is a required property",
+        ),
+        # missing `time`
+        (
+            {"time": None},
+            "time is a required property",
+        ),
+        # missing `data`
+        (
+            {"data": None},
+            "data is a required property",
+        ),
+        # missing `metadata`
+        (
+            {"metadata": None},
+            "metadata is a required property",
+        ),
+        # missing `jobId` in data
+        (
+            {"data": {"jobId": None}},
+            "jobId is a required property",
+        ),
+        # missing `jobStatus` in data
+        (
+            {"data": {"jobStatus": None}},
+            "jobStatus is a required property",
+        ),
+        # missing `correlationId` in metadata
+        (
+            {"metadata": {"correlationId": None}},
+            "correlationId is a required property",
+        ),
+        # invalid enum value for `jobStatus`
+        (
+            {"data": {"jobStatus": "INVALID_STATUS"}},
+            "data INVALID_STATUS is not one of [DESPATCHED, REJECTED]",
+        ),
+        # invalid `time` format
+        (
+            {"time": "invalid-time-format"},
+            "time invalid-time-format is not a date-time",
+        ),
+    ],
+)
+def test_process_letter_callback_validation_for_required_fields(
+    client, mock_dvla_callback_data, overrides, expected_error_message
+):
+    data = mock_dvla_callback_data(overrides=overrides)
+
+    response = client.post(url_for("notifications_letter_callback.process_letter_callback"), data=json.dumps(data))
+
+    response_json_data = response.get_json()
+    errors = response_json_data["errors"]
+
+    assert response.status_code == 400
+    assert any(
+        expected_error_message in error["message"] for error in errors
+    ), f"Expected error message '{expected_error_message}' not found in {errors}"
+
+
+@pytest.mark.parametrize(
+    "despatch_properties, expected_error_message",
+    [
+        # invalid enum for postageClass
+        (
+            [
+                {"key": "postageClass", "value": "invalid-postage-class"},
+                {"key": "totalSheets", "value": "5"},
+                {"key": "mailingProduct", "value": "MM UNSORTED"},
+                {"key": "Print Date", "value": "2024-08-01T09:15:14.456Z"},
+            ],
+            "data {key: postageClass, value: invalid-postage-class} is not valid under any of the given schemas",
+        ),
+        # missing totalSheets field
+        (
+            [
+                {"key": "postageClass", "value": "1ST"},
+                {"key": "mailingProduct", "value": "MM UNSORTED"},
+                {"key": "Print Date", "value": "2024-08-01T09:15:14.456Z"},
+            ],
+            "data [{key: postageClass, value: 1ST}, {key: mailingProduct, value: MM UNSORTED}, {key: Print Date, "
+            "value: 2024-08-01T09:15:14.456Z}] is too short",
+        ),
```

Comment:
should validation of `mailingProduct` also be included?

Reply:
https://github.com/alphagov/notifications-api/pull/4157/commits/912f29565533198d925088bc026697a73eb2ee13